[FLINK-36077][Connectors/Google PubSub] Implement table api support for SinkV2

vahmed-hamdy opened the pull request:
## Description
- Add Table Api support for PubSubSinkV2 [FLINK-36077]
- Added credential providers options for Json Credentials and Access token credentials
## Testing
- Added Unit tests
- Tested against GCP pubsub
## Documentation
- https://issues.apache.org/jira/browse/FLINK-36078

vahmed-hamdy commented:
@snuyanzin Could you please take a look?

vahmed-hamdy commented:
@snuyanzin Could you please review or nominate a reviewer please?

snuyanzin commented on flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/table/CredentialProviderFactory.java:
```diff
@@ -0,0 +1,120 @@
+            case GOOGLE_CREDENTIALS:
+                return defaultCredentialsProviderBuilder().build();
+            case FIXED:
+                String missingOptionsErrorMessage =
+                        "Fixed credential provider requires 'fixed.credentials-json' or 'fixed.credentials.access-token' "
+                                + "and 'fixed.credentials.access-expiration-epoch-millis' options to be set";
```
I guess error message could be built in lazy mode in corresponding method

snuyanzin commented on flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/table/CredentialProviderFactory.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    private GoogleCredentials getJsonCredentials(Map<String, String> options, String errorMessage) {
```
do we need input map here if only one option is checked?

snuyanzin commented on flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/table/CredentialProviderFactory.java:
```diff
@@ -0,0 +1,120 @@
+    private GoogleCredentials getAccessTokenCredentials(
+            Map<String, String> options, String errorMessage) {
+        Preconditions.checkNotNull(options, errorMessage);
+        String credentialsAccessToken = options.get("fixed.credentials.access-token");
```
can we have constants for such option names?

snuyanzin commented on flink-connector-gcp-pubsub/src/test/java/org/apache/flink/connector/gcp/pubsub/sink/table/CredentialProviderFactoryTest.java:
```diff
@@ -0,0 +1,202 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatExceptionOfType;
+
+/** Tests for {@link CredentialProviderFactory}. */
+public class CredentialProviderFactoryTest {
```
```suggestion
class CredentialProviderFactoryTest {
```
no need for `public`

snuyanzin commented on flink-connector-gcp-pubsub/src/test/java/org/apache/flink/connector/gcp/pubsub/sink/table/CredentialProviderFactoryTest.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    public void createDefaultCredentialsProvider() {
```
```suggestion
    void createDefaultCredentialsProvider() {
```
no need for `public` here and below

snuyanzin commented on flink-connector-gcp-pubsub/src/test/java/org/apache/flink/connector/gcp/pubsub/sink/table/PubSubDynamicSinkFactoryTest.java:
```diff
@@ -0,0 +1,296 @@
+import static org.apache.flink.table.factories.utils.FactoryMocks.createTableSink;
+
+/** Tests for {@link PubSubDynamicSinkFactory}. */
+public class PubSubDynamicSinkFactoryTest {
```
no need for public

snuyanzin commented on flink-connector-gcp-pubsub/src/test/java/org/apache/flink/connector/gcp/pubsub/sink/table/CredentialProviderFactoryTest.java:
```diff
@@ -0,0 +1,202 @@
+    @Test
+    public void createFixedCredentialsProviderFailsIfOptionsAreNull() {
+        CredentialProviderFactory factory = new CredentialProviderFactory();
+        assertThatExceptionOfType(NullPointerException.class)
```
I'm not sure NPE is a good exception if user is able to see it...
is it possible to use something like `IlleagalArgumentException` with corresponding error message?

snuyanzin commented on flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/table/CredentialProviderFactory.java:
```diff
@@ -0,0 +1,120 @@
+            return GoogleCredentials.fromStream(
+                    new ByteArrayInputStream(credentialsStringJson.getBytes()));
+        } catch (IOException e) {
+            throw new RuntimeException(
```
should it be also `IllegalArgumentException` ?

vahmed-hamdy commented:
@snuyanzin thanks for the review, I refactored the factory tests and removed unnecessary public access modifiers, It seems that the CI is failing, so I will take a look at that first.

snuyanzin commented:
>CI is failing,
 
i guess the reason is dependency convergence since at some point commons-codec in Flink main repo was bumped to 1.16.1
may be just use explicit dependency in pom in this repo rather than relying on the one coming from Flink

vahmed-hamdy commented:
@snuyanzin yeah, good idea. Fixed now
